### ISSUE TITLE
[FR] Slow array function used in loop: count()

### ISSUE THREAD

**Issue by orklah** [kalessil/phpinspectionsea#1280]
Description (screenshot):
--
I would like to suggest to suggest to replace usage of count() in loops. It could be added in "Slow array function used in loop" or in a separate inspection.

The goal is to match cases like these:
```php
function scope(){
    $prime_numbers = [1,3,5,7];
    foreach($prime_numbers as $i){
        if($i > 3 && count($prime_numbers) > 3){
            echo 'test';
        }
    }
}
```

It should work with all loops, not only foreach. We will want to check if there is no write access to the counted variable in the loop to suggest to move the count() in a local variable out of the loop.

Note: there is already an inspection that could warn if count() usage doesn't seem related to the loop but it works only with foreach and it is not reported in my example because there is $i in the if()

Thanks!

**Comment by dryabov** [comment]
The ideal solution would be to introduce `@pure` attribute in phpDoc, properly annotate all built-in functions (including `\count`), and add inspections that detect
1) pure function calls inside a loop with arguments that are not changed in the loop (with suggestion to move such a call outside of the loop),
2) subsequent pure function calls with arguments that are not changed between calls (with suggestion to save result of the former call).

**Comment by orklah** [comment]
> The ideal solution would be to introduce `@pure` attribute in phpDoc, properly annotate all built-in functions (including `\count`), and add inspections that detect
> 
> 1. pure function calls inside a loop with arguments that are not changed in the loop (with suggestion to move such a call outside of the loop),
> 2. subsequent pure function calls with arguments that are not changed between calls (with suggestion to save result of the former call).

We would need to convince jetbrains to update their stubs and I'm not sure they would accept without some pre-existing community usage.
If the stubs aren't changed, nobody will bother to do it for a simple inspection...

This would be optimal but too much a hassle

**Comment by dryabov** [comment]
OK, then there is another simple way to identify function as pure one: it is a function with arguments that are passed by value and are not objects. Of course, in this case we will not able to detect pure functions with object arguments or pure class methods, but it should cover most built-in functions (with possibility to introduce a white-list for other built-in functions/methods).

**Comment by kalessil** [comment]
Most of the core functions are not causing, problems (if we are talking about the practice field).
The `count` is one of problematic (even op-cached) + from my past optimization fun I remember the `array_unique` function - will add them both as slow.

And apparently I need to re-evaluate cases where subject variables are changed in a loop, I totally missed that part.
